### PR TITLE
Fixing build issues

### PR DESCRIPTION
**Pull request by syclik**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Fixes a lot of build issues.

This allows users in Windows, Linux, and Mac to run CmdStan without additional configuration. The defaults allow models to be built without additional configuration.

#### Intended Effect:

Fixes build issues that users were reporting:
- #586 
- #590 
- #588 

#### How to Verify:

Build from:
- Mac, Windows, and Linux
- from a fresh clone without a `make/local` or `~/.config/cmdstan/make.local` file present

And build the bernoulli example.

Please verify that this works on your local machine or whatever other machine you have access to.

#### Side Effects:

These are the known side-effects in this pull request. These shouldn't affect users. The second one may affect developers:
- Mac: on develop, we have included in the `CXXFLAGS` `-stdlib=libc++`. In this branch, this is omitted. 
    Without it, it builds fine on my machine. I removed the default compiler from being set to `clang++`. Instead, we use the system default, which usually defaults to `c++`. The math library's `detect_cc` searches for the string `clang++` and uses that to set options.
- Linux: tests don't build with the default values. In order to build them, you need to include `GTEST_CXXFLAGS += -DGTEST_HAS_PTHREAD=0` in `make/local` or `~/.config/cmdstan/make.local`.
    This is also happening because `CXX`'s default value is `c++`. The `os_linux` file uses the value there (indirectly) to add to the `GTEST_CXXFLAGS`. We should probably just add that directly for all of Linux, but the fix should be in the Math library. This has been documented.

#### Documentation:

- Added subsection mentioning how to build tests in linux. 
- Minor updates to change `CC` to `CXX` (this may conflict with the other PR, but we can work through that easily).

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Generable

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by seantalts** [comment]
Hey, I think there's a lot of confusion going on about what build issues actually exist - check out the post here: https://github.com/stan-dev/cmdstan/issues/590#issuecomment-351215438 

Maybe we can hop on a hangout or something? I don't actually think there are any build issues at this point (now that we finally merged that PR to fix the pch thing for g++ people), assuming we doc correctly how to set CC in make/local.

Also, I think it's probably a good idea to switch to `g++` as the default compiler everywhere (I'm setting up a PR in math repo) because that's aliased on macs to clang++, so it will work out of the box for many more people.

**Comment by syclik** [comment]
Sure, we can jump on a hangout.

Re: build issues. We have a few with v2.17.0. What I expect is in mac,
linux, and windows, with a normal-ish setup, it would run out of the box.
Currently, that doesn't happen for linux and windows. That's what I was
trying to fix here.

File: src/docs/cmdstan-guide/installation.tex [hunk 1]
```diff
@@ -454,6 +454,20 @@ \subsection{Downloading and Unpacking \CmdStan Source}
...
 \end{quote}
 
+\subsection{Running \CmdStan Tests}
+
+Google test uses pthread by default. If the library is not available on your
+system, it's still easy to run the tests. Add this to your \code{make/local}
+file:
+\begin{quote}
+\begin{Verbatim}[fontshape=sl,fontsize=\small]
+GTEST_CXXFLAGS += -DGTEST_HAS_PTHREAD=0
```
Comment (seantalts) [inline src/docs/cmdstan-guide/installation.tex]: why is this step suddenly necessary? Does Jenkins have to change?

Comment (syclik) [reply src/docs/cmdstan-guide/installation.tex]: Just hunted it down. It has to do with the os_linux in the math library setting it for pretty much every compiler option there. It trickles down from `CC`. Let me figure out a different way around it.

**Comment by syclik** [comment]
Issues in other repos for CXX / CC:
- math already exists: stan-dev/math#617
- stan (created a new issue): stan-dev/stan#2444

**Comment by syclik** [comment]
@seantalts, just made some changes. To address the three things:
- There are now issues in Stan and Math for using CXX; I didn't add one for CmdStan because it's fine with CXX now.
- I added a comment to the very top of the makefile. Feel free to change or make suggestions.
- for the installation.tex, I didn't change it just because it's described in the subsection right before. If you have any suggestions on how to make that better, want to do that?

**Comment by syclik** [comment]
I also updated `make/tests` to include the gtest-related pthread flag for linux. I left a fixme so we remember to remove it later.

**Pull request merged**